src/explorer/Common/ValueWithCountList.cs

```diff
@@ -22,47 +22,51 @@ public void AddValueCount(T value, long count)
             Add((value, TotalCount + count));
         }
 
-        public T GetRandomValue(Random rand, T @default)
+        public T GetRandomValue(Random rand)
         {
-            if (Count == 0)
-            {
-                return @default;
-            }
             var rcount = rand.NextLong(TotalCount);
-            return FindSubstring(rcount);
+            return FindValue(rcount);
         }
 
-        private T FindSubstring(long count)
+        public T FindValue(long count)
         {
+            if (TotalCount == 0)
+            {
+                throw new InvalidOperationException("Collection is empty.");
+            }
+            if (count < 0 || count >= TotalCount)
+            {
+                throw new ArgumentException($"The {nameof(count)} parameter should have a value between 0 and {TotalCount - 1}, inclusive.");
+            }
             var left = 0;
             var right = Count - 1;
             while (true)
             {
                 var middle = (left + right) / 2;
-                if (middle == 0 || middle == Count - 1)
-                {
-                    return this[middle].Value;
-                }
                 if (count < this[middle].Count)
                 {
+                    if (middle == left)
+                    {
+                        return this[middle].Value;
+                    }
                     if (count >= this[middle - 1].Count)
                     {
-                        return this[middle - 1].Value;
+                        return this[middle].Value;
                     }
                     right = middle;
                 }
-                else if (count > this[middle].Count)
+                else if (count >= this[middle].Count)
                 {
-                    if (count <= this[middle + 1].Count)
+                    if (middle == right)
                     {
                         return this[middle].Value;
                     }
+                    if (count < this[middle + 1].Count)
+                    {
+                        return this[middle + 1].Value;
```

Comment:
Shouldn't this be:

```suggestion
                        return this[middle].Value;
```

(?)

Reply:
I checked the code and the tests again and this looks good to me. Actually, the condition above this
```                    
if (middle == right)
{
    return this[middle].Value;
}
```
seems to be unnecessary.

Reply:
I don't really understand this. If `count < this[middle + 1].Count` is true, then surely this excludes the value at `[middle + 1]` since the `<` is non-inclusive? 

I would keep the `middle == right` condition - it seems to make sense, and I can think of scenarios where it might be triggered. Maybe the test cases are not exhaustive? I noticed the test cases are all evenly weighted - I think some of the edge cases would be triggered with non-even weighting. 

Let's merge this for now and we can try to write some better tests cases afterwards.